# Hyperparameter Tuning - Iris

In [18]:
import warnings
warnings.filterwarnings('ignore')
from sklearn import datasets
from sklearn.svm import SVC
iris = datasets.load_iris()

In [19]:
iris.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [20]:
import pandas as pd
import numpy as np

df = pd.DataFrame(iris.data, columns=iris.feature_names)

In [21]:
df['flower'] = pd.DataFrame(iris.target, columns=['flower'])

In [22]:
df['flower'] = df['flower'].apply(lambda x: iris.target_names[x])

In [23]:
x = df.drop(['flower'],axis=1)
y = df['flower']

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.30)

In [25]:
model = SVC(kernel='rbf',C=30, gamma ='auto')
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.9555555555555556

In [26]:
from sklearn.model_selection import cross_val_score


In [30]:
cross_val_score(SVC(kernel='rbf',C=10, gamma ='auto'),iris.data, iris.target, cv=5)

array([0.96666667, 1.        , 0.96666667, 0.96666667, 1.        ])

In [31]:
cross_val_score(SVC(kernel='linear',C=10, gamma ='auto'),iris.data, iris.target, cv=5)

array([1.        , 1.        , 0.9       , 0.96666667, 1.        ])

In [32]:
cross_val_score(SVC(kernel='rbf',C=20, gamma ='auto'),iris.data, iris.target, cv=5)

array([0.96666667, 1.        , 0.9       , 0.96666667, 1.        ])

In [36]:
from sklearn.model_selection import GridSearchCV

clf = GridSearchCV(SVC(gamma='auto'),{
    'C':[1,10,20],
    'kernel':['rbf','linear']
}, cv=5, return_train_score=False)

clf.fit(iris.data,iris.target)
clf.cv_results_

{'mean_fit_time': array([0.0021997 , 0.00179906, 0.00299516, 0.00119953, 0.00139937,
        0.00100126]),
 'std_fit_time': array([3.99565821e-04, 4.00043823e-04, 1.09695285e-03, 4.00758538e-04,
        4.88636760e-04, 2.94946043e-06]),
 'mean_score_time': array([0.00119977, 0.00139961, 0.00179758, 0.00099998, 0.00120029,
        0.00059986]),
 'std_score_time': array([3.99446729e-04, 4.89415322e-04, 7.44230255e-04, 2.27386837e-06,
        4.00734884e-04, 4.89784675e-04]),
 'param_C': masked_array(data=[1, 1, 10, 10, 20, 20],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 10, 'kernel': 'rbf'},
  {'C': 10, 'kernel': 'linear'},
  {'C': 20, 'ker

In [37]:
result = pd.DataFrame(clf.cv_results_)

In [39]:
result

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       0.002200      0.000400         0.001200        0.000399       1   
1       0.001799      0.000400         0.001400        0.000489       1   
2       0.002995      0.001097         0.001798        0.000744      10   
3       0.001200      0.000401         0.001000        0.000002      10   
4       0.001399      0.000489         0.001200        0.000401      20   
5       0.001001      0.000003         0.000600        0.000490      20   

  param_kernel                         params  split0_test_score  \
0          rbf      {'C': 1, 'kernel': 'rbf'}           0.966667   
1       linear   {'C': 1, 'kernel': 'linear'}           0.966667   
2          rbf     {'C': 10, 'kernel': 'rbf'}           0.966667   
3       linear  {'C': 10, 'kernel': 'linear'}           1.000000   
4          rbf     {'C': 20, 'kernel': 'rbf'}           0.966667   
5       linear  {'C': 20, 'kernel': 'linear'}           1.000000   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0                1.0           0.966667           0.966667                1.0   
1                1.0           0.966667           0.966667                1.0   
2                1.0           0.966667           0.966667                1.0   
3                1.0           0.900000           0.966667                1.0   
4                1.0           0.900000           0.966667                1.0   
5                1.0           0.900000           0.933333                1.0   

   mean_test_score  std_test_score  rank_test_score  
0         0.980000        0.016330                1  
1         0.980000        0.016330                1  
2         0.980000        0.016330                1  
3         0.973333        0.038873                4  
4         0.966667        0.036515                5  
5         0.966667        0.042164                6

In [41]:
result[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0       1          rbf         0.980000
1       1       linear         0.980000
2      10          rbf         0.980000
3      10       linear         0.973333
4      20          rbf         0.966667
5      20       linear         0.966667

In [42]:
clf.best_score_

0.9800000000000001

In [43]:
clf.best_params_

{'C': 1, 'kernel': 'rbf'}

In [48]:
from sklearn.model_selection import RandomizedSearchCV
rs = RandomizedSearchCV(SVC(gamma='auto'),{
    'C':[1,10,20],
    'kernel':['rbf','linear']
}, cv=5, return_train_score=False, n_iter=2)

rs.fit(iris.data, iris.target)

pd.DataFrame(rs.cv_results_)[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0      20       linear         0.966667
1      10       linear         0.973333

# CHOOSING THE BEST MODEL 

In [49]:
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [61]:
model_params = {
                'svm':{
                    'model':SVC(gamma='auto'), 'params':{
                                                    'C':[1,10,20],
                                                    'kernel':['rbf','linear']
                    }
                },
                'random_forest':{
                    'model':RandomForestClassifier(), 'params':{
                                                    'n_estimators':[1,5,10]
                    }
                },
                'Logistic_Regression':{
                    'model':LogisticRegression(solver='liblinear', multi_class='auto'), 'params':{
                                                    'C':[1,5,20]
                    }
                }
}

In [64]:
scores = []

for model_name, mp in model_params.items():
    clf = GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(iris.data, iris.target)
    scores.append({
        'model':model_name,
        'best_score':clf.best_score_,
        'best_params':clf.best_params_
    })

In [65]:
df = pd.DataFrame(scores, columns=['model','best_score','best_params'])
df

model  best_score                best_params
0                  svm    0.980000  {'C': 1, 'kernel': 'rbf'}
1        random_forest    0.953333       {'n_estimators': 10}
2  Logistic_Regression    0.966667                   {'C': 5}

We see that for the iris data set the best model among the used models is SVM with a score of 0.98